### DISCUSSION TITLE
Combining Retext with Remark / markdown-magic as a Shared Config

### DISCUSSION THREAD

**Discussion by ProfessorManhattan** [remarkjs/remark#1137] [Q&A] [answered]
Hey, I'm basically trying to make Retext and Remark run at the same time. My config looks something like this:

```
const remarkPreset = {
  plugins: [
    remarkParse,
    remarkToc,
    [remarkRetext, unified().use(retextEquality)]
  ]
}
export default remarkPreset
```

~~It works until I add the `remarkRetext` array (3rd plugin). But, that's giving me an error that looks like this:~~

It looks like its running without error if I add `retextEnglish` first. However, no lint messages are being output.

I'd like to combine all this stuff into a single shared config because I already have like 8 different shared configs for various tools. I also thought it might be more performant if Remark only had to read / parse the files once.

Does anybody know how I can include retext plugins inside of a remark-preset? Is there perhaps a way I can get the stringified version of the markdown and run some synchronous code on it before passing the formatted data back into remark?

**Comment by ChristianMurphy** [comment, marked as answer]
Heya @ProfessorManhattan! 👋 

It works as is, here's an example: https://codesandbox.io/p/sandbox/nifty-sky-f8lp7l?selection=%5B%7B%22endColumn%22%3A43%2C%22endLineNumber%22%3A1%2C%22startColumn%22%3A43%2C%22startLineNumber%22%3A1%7D%5D&file=%2FREADME.md
If you're trying something different please share a complete example, ideally in a sandbox.

Some ideas of what could be going on.
1. you are adding a custom `remark-parse`, overriding the existing parser, avoid doing so unless you need to customize the parser.
2. make sure you have your module as `type: module` in package.json, or your file has the `.mjs` extension.
3. make sure you select the language you want retext to parse with, likely `retext-english`
4. make sure your dependencies are running the latest major version

**Reply by ProfessorManhattan** [reply]
Hey @ChristianMurphy -- thanks for that. That's how my code is and it's not reporting `retext` issues. All my deps are latest. And I just loaded your link but it's not working on the site. Here's the config I'm using: https://gitlab.com/megabyte-labs/npm/configs/remark-preset-strict-lint/-/blob/master/src/main.ts

**Reply by ChristianMurphy** [reply]
> And I just loaded your link but it's not working on the site

In codesandbox, in the right panel, run the test task.
![Screenshot 2023-03-23 at 15-09-32 remark-retext - CodeSandbox](https://user-images.githubusercontent.com/3107513/227376000-6bc3fcde-c0eb-4049-b16f-5c4aa07af489.png)

> That's how my code is 

It is not.
It overrides the remark parser provided by the remark-cli in https://gitlab.com/megabyte-labs/npm/configs/remark-preset-strict-lint/-/blob/master/src/main.ts#L128
It sets two different retext parsers https://gitlab.com/megabyte-labs/npm/configs/remark-preset-strict-lint/-/blob/master/src/main.ts#L99-100
And there is a bunch of abstraction for enabling and disabling rules, which could be interferring.

The file is typescript, and it is unclear if the config will support ESM on Node https://gitlab.com/megabyte-labs/npm/configs/remark-preset-strict-lint/-/blob/master/tsconfig.json
For more consistent ESM compat use `module: node16`

https://github.com/remarkjs/remark/blob/e456dc5b0fb621dfa5f1ecae455812c6e65a25d0/tsconfig.json#L6

---

TL;DR the sandbox is a working example of how it can be used.
You can start from there if you want.
Or you boil your configuration down to a minimal reproducible example https://stackoverflow.com/help/minimal-reproducible-example, the process will help isolate what is actually the root cause.

**Reply by ProfessorManhattan** [reply]
Hey @ChristianMurphy -- thanks since you confirmed that it was possible, that was enough for me to give another go and I found that `remark-copy-linked-files` (which actually might not be working anyway) had to be after the inclusion of the `unified()` retext plugin references. It works now. TY!!

**Reply by ProfessorManhattan** [reply]
It is working with both `retext-english` and `retext-latin` --- I might just remove `retext-latin` though... I was under the impression that it had something to do with encoding. I'm also under the impression that latin has not been used as an actual language in the past few thousand years. 😀

**Comment by ProfessorManhattan** [comment]
Thanks
